### PR TITLE
Fix for issue #1558 - Search key mappings in Freon

### PR DESCRIPTION
**Pull request by cribalik**
croutonfbserver now sends information to kiwi nacl module on startup regarding whether we are using Freon or not. The nacl module uses different key schemes depending on whether Freon is used or not, enabling SearchKey + otherkey mappings to work correctly for both freon and non-freon users.

### CODE REVIEW

File: host-ext/nacl_src/kiwi.cc [hunk 1]
```diff
@@ -246,6 +246,12 @@ class KiwiInstance : public pp::Instance {
         return false;
     }
 
+    /* Recieves and processes initialization information */
+    void SocketParseInitInformation(const char* data, int datalen) {
```
Comment (dnschneid) [inline host-ext/nacl_src/kiwi.cc]: Is the idea that later revs would just make datalen longer, so you don't need to bother with versioning?

File: src/fbserver.c [hunk 2]
```diff
@@ -491,6 +491,19 @@ int write_cursor() {
     return 0;
 }
 
+void write_init() {
+    struct stat _stat;
```
Comment (dnschneid) [inline src/fbserver.c]: Is there a better way to name this variable, other than a leading _?

**Comment by dnschneid** [comment]
This looks quite nice, thanks!  @drinkcat, any thoughts?

File: src/fbserver.c (hunk 2 above)
Comment (cribalik) [inline src/fbserver.c]: Well it's a buffer for the `stat` call.. so how about `stat_buf`?

File: host-ext/nacl_src/kiwi.cc (hunk 1 above)
Comment (cribalik) [inline host-ext/nacl_src/kiwi.cc]: I think that might present problems (adding a field to the `initinfo` struct doesn't necessarily increase `datalen`). 
I was thinking of using the server_version_ variable for versioning instead.

**Comment by dnschneid** [comment]
To be clear, this is compatible with your other PR (#1832), right?

File: src/fbserver.c [hunk 3]
```diff
@@ -491,6 +491,19 @@ int write_cursor() {
     return 0;
 }
 
+void write_init() {
+    struct stat stat_buf;
+    char raw[FRAMEMAXHEADERSIZE + sizeof(struct initinfo)];
+    struct initinfo* i = (struct initinfo*)(raw + FRAMEMAXHEADERSIZE);
+    i->type = 'I';
+    i->freon = 0;
+    if (stat("/sys/class/tty/tty0/active", &stat_buf) == -1) {
```
Comment (drinkcat) [inline src/fbserver.c]: Using `access` is arguably a bit simpler...

**Comment by drinkcat** [comment]
This looks awesome! I'm happy as it is, using `access` instead of `stat` makes `write_init` a tiny bit simpler, but, really, up to you.

**Comment by cribalik** [comment]
Good call @drinkcat, changed it to `access`.

**Comment by cribalik** [comment]
@dnschneid Yes, this PR is compatible and non-redundant with my other PR (#1832).

File: host-ext/nacl_src/kiwi.cc [hunk 4]
```diff
@@ -257,7 +267,7 @@ class KiwiInstance : public pp::Instance {
 
         if (server_version_ != VERSION) {
             /* TODO: Remove VF1 compatiblity */
-            if (server_version_ == "VF1") {
+            if (server_version_ == "VF1" || server_version_ == "VF2") {
```
Comment (dnschneid) [inline host-ext/nacl_src/kiwi.cc]: Actually, can we not pop up the warning for the update?  Most people probably won't care in this case...

Comment (dnschneid) [inline host-ext/nacl_src/kiwi.cc]: Eh, I take that back.  Maybe it'll force people to update occasionally :)

**Pull request merged**